tests/components/edl21/test_config_flow.py

```diff
@@ -0,0 +1,106 @@
+"""Test EDL21 config flow."""
+
+import pytest
+
+from homeassistant import data_entry_flow
+from homeassistant.components.edl21.const import CONF_SERIAL_PORT, DEFAULT_NAME, DOMAIN
+from homeassistant.config_entries import SOURCE_IMPORT, SOURCE_USER
+from homeassistant.const import CONF_NAME
+from homeassistant.core import HomeAssistant
+
+from tests.common import MockConfigEntry
+
+VALID_CONFIG = {CONF_NAME: "My Smart Meter", CONF_SERIAL_PORT: "/dev/ttyUSB1"}
+
+pytestmark = pytest.mark.usefixtures("mock_setup_entry")
+
+
+async def test_show_form(hass: HomeAssistant) -> None:
+    """Test that the form is served with no input."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.FlowResultType.FORM
+    assert result["step_id"] == SOURCE_USER
+
+
+async def test_integration_already_exists(hass: HomeAssistant) -> None:
+    """Test that a new entry must not have the same name or serial port as an existing entry."""
+
+    MockConfigEntry(
+        domain=DOMAIN,
+        data=VALID_CONFIG,
+    ).add_to_hass(hass)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_USER},
+        data=VALID_CONFIG,
+    )
+
+    assert result["type"] == data_entry_flow.FlowResultType.ABORT
+    assert result["reason"] == "already_configured"
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_USER},
+        data={
+            CONF_NAME: "Other Smart Meter",
+            CONF_SERIAL_PORT: VALID_CONFIG[CONF_SERIAL_PORT],
+        },
+    )
+
+    assert result["type"] == data_entry_flow.FlowResultType.ABORT
+    assert result["reason"] == "already_configured"
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_USER},
+        data={CONF_NAME: VALID_CONFIG[CONF_NAME], CONF_SERIAL_PORT: "/dev/ttyUSB2"},
+    )
+
+    assert result["type"] == data_entry_flow.FlowResultType.ABORT
+    assert result["reason"] == "already_configured"
+
+
+async def test_create_entry_by_user(hass: HomeAssistant) -> None:
```

Comment:
This test should really be merged with `test_show_form` to show end-to-end (`async_init` with context and no data, then `async_configure` with the input)